src/DocumentationViewer.js

```diff
@@ -0,0 +1,219 @@
+  const forward = builder.get_object("forward");
+
+  const base_path = "/app/share/doc";
+
+  webview.load_uri(`file://${base_path}/gtk4/index.html`);
+  container.child = webview;
+  let loaded = false;
+
+  webview.connect("load-changed", (view, load_event) => {
+    if (load_event === WebKit.LoadEvent.FINISHED) {
+      loaded = true;
+      sidebar_button.active = false;
+    } else {
+      loaded = false;
+    }
+  });
+
+  back.connect("clicked", () => {
+    webview.go_back();
+  });
+
+  forward.connect("clicked", () => {
+    webview.go_forward();
+  });
+
+  sidebar_button.connect("toggled", () => {
+    if (loaded) {
+      if (sidebar_button.active) disableDocSidebar(webview);
+      else enableDocSidebar(webview);
+    }
+  });
+
+  sidebar_button.connect("toggled", () => {
+    if (loaded) {
+      if (sidebar_button.active) disableDocSidebar(webview);
+      else enableDocSidebar(webview);
+    }
+  });
+
+  search_entry.connect("search-changed", () => {
+    listbox.invalidate_filter();
+  });
+
+  search_bar.bind_property(
+    "search-mode-enabled",
+    search_button,
+    "active",
+    GObject.BindingFlags.BIDIRECTIONAL,
+  );
+
+  search_button.connect("toggled", () => {
+    if (search_button.active) sidebar_button.active = true;
+  });
+
+  listbox.set_sort_func(sort);
+  listbox.invalidate_sort();
+  listbox.set_filter_func(filter);
+
+  getDocs(base_path)
+    .then((docs) => {
+      for (const doc in docs) {
+        const row = new Adw.ActionRow({
+          title: doc,
+        });
+        row.add_suffix(new Gtk.Image({ icon_name: "go-next-symbolic" }));
+        listbox.append(row);
+      }
+      listbox.connect("row-selected", (_, row) => {
+        webview.load_uri(docs[row.title]);
+      });
+      return docs;
+    })
+    .catch(logError)
+
+
+
+  function sort(row1, row2) {
+    return row1.title > row2.title;
+  }
+
+  function filter(row) {
+    const re = new RegExp(search_entry.text, "i");
+    return re.test(row.title);
+  }
+
+  const action_documentation = new Gio.SimpleAction({
+    name: "documentation",
+    parameter_type: null,
+  });
+  action_documentation.connect("activate", () => {
+    window.present();
+  });
+  application.add_action(action_documentation);
+  application.set_accels_for_action("app.documentation", ["<Control><Shift>D"]);
+}
+
+async function getDocs(base_path) {
+  try {
+    const docs = {};
+    const dirs = await list(base_path);
+
+    for (const dir of dirs) {
+      try {
+        const index = await readDocIndex(base_path, dir);
+        const namespace = `${index["meta"]["ns"]}-${index["meta"]["version"]}`;
+        const path = `file://${base_path}/${dir}/index.html`;
+        docs[namespace] = path;
+      } catch (e) {
+        // Ignore the error if the dir does not contain index.json
+        if (!e.matches(Gio.IOErrorEnum, Gio.IOErrorEnum.NOT_FOUND)) logError(e);
+      }
+    }
+
+    return docs;
+  } catch (e) {
+    logError(e);
+  }
```

Comment:
since you have a error handler up the chain in https://github.com/sonnyp/Workbench/pull/358/files#diff-6533bf97682b99bf434791a454793569a69668f437288ddcedac958454abf4a7R122, you can remove the try/catch in this function